### ISSUE TITLE
MaxUnpool indices incompatible with PyTorch and inconsistent with MaxPool for kernels sizes which aren't a multiple of output size

### ISSUE THREAD

**Issue by fdwr** [onnx/onnx#2398]
## Summary
ONNX's [`MaxUnpool-11`](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#MaxUnpool-11) indexing is incompatible with PyTorch's MaxUnpool for kernel sizes which aren't a multiple of the output size. ONNX MaxUnpool is even incompatible with ONNX's own [`MaxPool-11`](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#MaxPool-11) for such cases, as MaxPool outputs indices as a large 1D tensor agnostic to padding or kernel size/stride (consistent with PyTorch) whereas MaxUnpool seems to be doing something weird related to the inferred output shape and ignoring the explicitly specified `output_shape`. So feeding MaxPool's indices output onward to MaxUnpool gives gibberish results since they map to the wrong element offsets (or worse, [fail to execute the model](https://github.com/microsoft/onnxruntime/issues/2154)).

The backend example `with_output_shape` should fix indices `xI` to the following:

``` python
node = onnx.helper.make_node(
    'MaxUnpool',
    inputs=['xT', 'xI', 'output_shape'],
    outputs=['y'],
    kernel_shape=[2, 2],
    strides=[2, 2]
)
xT = np.array([[[[5, 6],
                 [7, 8]]]], dtype=np.float32)
# From:
# xI = np.array([[[[5, 7],
#                 [13, 15]]]], dtype=np.int64)
# To:
xI = np.array([[[[6, 8], # Corrected values
                 [16, 18]]]], dtype=np.int64)
output_shape = np.array((1, 1, 5, 5), dtype=np.int64)
y = np.array([[[[0, 0, 0, 0, 0],
                [0, 5, 0, 6, 0],
                [0, 0, 0, 0, 0],
                [0, 7, 0, 8, 0],
                [0, 0, 0, 0, 0]]]], dtype=np.float32)
expect(node, inputs=[xT, xI, output_shape], outputs=[y], name='test_maxunpool_export_with_output_shape')
```

The indices given in the example only make sense if `output_shape` is 4x4.

## Further explanation
[`MaxPool-11`](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#MaxPool-11) finds maximal values in the unpooled input tensor and outputs indices of the maximal values selected, where indices are 0-based offsets pointing into the input tensor as if it was a huge 1D tensor (per spec, padding is irrelevant to this index value) with index values ranging `[0, N x C x D1 x ... x Dn)`. "The indices are computed as flatten 1-D tensor, and the indices do not consider padding."

[`MaxUnpool-11`](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#MaxUnpool-11) then reads those maximal values and indices (tensor output from `MaxPool`), restoring the maximal values to their element offsets in the unpooled output tensor (which has the same size and type as the corresponding original input tensor to `MaxPool`) and writing zeros to all other values in the output.

These two operators are roughly symmetric in that the output of one reverses the other. So calling `MaxUnpool(MaxPool(...))` yields the original input tensor (for the maximal elements anyway, leaving other elements zero), and the reverse composition `MaxPool(MaxUnpool(...))` returns output identical to the input tensor.

For pooling cases where the kernel evenly fits into the MaxUnpool output, there is no discrepancy between ONNX MaxPool/MaxUnpool (matches PyTorch). For odd sizes where the kernel doesn't even fit evenly, and the inferred shape differs from the actual full shape (e.g. 5x5 with 2x2 kernel and 2x2 strides), ONNX MaxPool is compatible with PyTorch (and yields the expected indices), but ONNX MaxUnpool is not - even when you pass the correct `output_shape` to MaxUnpool, since it appears ONNX MaxUnpool is bewilderingly using the inferred output shape (4x4) for indexing rather than the actual shape passed to it (5x5). So, values end up being written back to the wrong elements.

Now, there _is_ a potential ambiguity with MaxUnpool's output tensor size since more than one possible input tensor size to MaxPool reduces to the same output tensor size (e.g. MaxPool maps both a 2x2 and a 3x3 input tensor to 1x1 output, when given a 2x2 kernel and 2x2 stride), but all the examples below use `output_shape` explicitly, and so this is not the issue - the issue is the indexing.

## Example case
The original example is a 5x5 input, but we can simplify the example even more to use 3x3 input with 2x2 kernel and 2x2 strides:

### Code

``` python
import torch;

pool = torch.nn.MaxPool2d(kernel_size=2, stride=2, return_indices=True)
unpool = torch.nn.MaxUnpool2d(kernel_size=2, stride=2)
    
original_input = torch.tensor(
                    [[[[0, 0, 0],
                       [3, 0, 0],
                       [0, 0, 0]]]],
                    dtype=torch.float32
                    );
    
pooled_values, indices = pool(original_input)
unpooled = unpool(pooled_values, indices, output_size=torch.Size([1, 1, 3, 3]))

print("original_input:\n", original_input, sep='')
print("maxpool_output:\n", pooled_values, sep='')
print("maxpool_indices:\n", indices, sep='')
print("maxunpool_output:\n", unpooled, sep='')
print()
```

``` json
    {
      "op_type": "MaxPool",
      "version": 8,
      "kernel_shape": [2, 2],
      "pads": [0,0, 0,0],
      "strides": [2,2],
      "storage_order": 0,
      "X":       [[[[ 0, 0, 0],
                    [ 3, 0, 0],
                    [ 0, 0, 0]]]],
      "Y":       [[[[ 3]]]],
      "indices": [[[[ 3]]]],
      "T": "float32"
    },
    {
      "op_type": "MaxUnpool",
      "version": 9,
      "kernel_shape": [2, 2],
      "strides": [2, 2],
      "X": [[[[3]]]],
      "I": [[[[ 3]]]],
      "output_shape": [1, 1, 3, 3],
      "Y": [[[[0, 0, 0],
              [3, 0, 0], // <- expected
              [0, 0, 0]]]],
      "T1": "float32"
    }
```

### Output

PyTorch output:
``` python
tensor([[[[0., 0., 0.], # original values (MaxPool input)
          [3., 0., 0.],
          [0., 0., 0.]]]])
tensor([[[[3.]]]]) # maximal value
tensor([[[[3]]]]) # indices output
tensor([[[[0., 0., 0.], # MaxUnpool output (matches MaxPool input)
          [3., 0., 0.],
          [0., 0., 0.]]]])
```

ONNX output:
``` python
tensor([[[[0., 0., 0.], # original values (MaxPool input)
          [3., 0., 0.],
          [0., 0., 0.]]]])
tensor([[[[3.]]]]) # maximal value
tensor([[[[3]]]]) # indices output
tensor([[[[0., 0., 0.], # MaxUnpool output (should match MaxPool input)
          [0., 3., 0.], <-- mismatched!
          [0., 0., 0.]]]])
```

## Proposed fix 
Make MaxUnpool use the same indexing as MaxPool, just like how MaxPool treats the indices purely as offsets within the unpooled tensor (regardless of padding or strides or kernel size). *The reverse alternative would be to make MaxPool's indexing compatible with MaxUnpool, but then ONNX MaxPool would be incompatible with PyTorch which hampers interop.

**Comment by eudora-jia** [comment]
what does the expect mean? 
expect(node, inputs=[xT, xI, output_shape], outputs=[y], name='test_maxunpool_export_with_output_shape')
 @eudora-jia
 
 

what does the expect mean? @huitseeker 

Attach files by dragging & dropping, selecting or pasting them.
Assignees
No one assigned
Labels
None yet
Projects
None yet
Milestone
No milestone
Notifications
Customize
You’re not receiving notifications from this thread.
1 participant
@fdwr
@huitseeker

**Comment by fdwr** [comment]
I tried comparing against TensorFlow, but I didn't see an *unpool* equivalent. MaxPool does have the same indexing though as PyTorch (agnostic to kernel size and padding).

``` python
import tensorflow as tf

input = tf.constant([[[[0.0],[0.0],[0.0]],[[3.0],[0.0],[0.0]],[[0.0],[0.0],[0.0]]]])
ksize = (1,2,2,1)
strides = (1,2,2,1)

maxpool = tf.nn.max_pool_with_argmax(input, ksize, strides, "VALID", 'NHWC')
# No max unpool equivalent?

sess = tf.Session()
result = sess.run(maxpool)

print("MaxPool value, indices", result.output, result.argmax)
```

**Comment by skottmckay** [comment]
The description of output_shape probably causes confusion

> output_shape (optional) : T2
> The shape of the output can be explicitly set which will cause pads values to be auto generated. If 'output_shape' is specified, 'pads' values are ignored.

And seems counter to the op description which says

> The third input argument, output_size, is meant to disambiguate the op and produce output tensor of known/predictable size.

If the purpose of output_shape is to disambiguate, it shouldn't be trying to auto generate pads should it?

**Comment by snnn** [comment]
Should it be addressed in the new release? @faxu @hinhuang007

**Comment by chinhuang007** [comment]
@snnn I assume the issue is not fixed by PR 2193. Then I would suggest to put this as a post release 1.7 fix to minimize the release dependency.

**Comment by skottmckay** [comment]
This still needs to be clarified.

**Comment by fdwr** [comment]
(minor epiphany) `MaxUnpool` is really just `ConstantOfShape` and `ScatterElements`, So after computing the shapes, the functional decomposition of...

```diff
MaxUnpool(
    kernel_shape,
    pads,
    strides,
-   X, // poorly named - "input" is clearer and symmetric with "output"
+   input
-   I, // poorly named and inconsistent with MaxPool which uses *indices* like ScatterElements.
+   indices
    output_shape,
    output
)
```

...is really...

```
ScatterElements(
    input = ConstantOfShape(value = 0, shape = outputShape),
    indices = Reshape(indices, flattenedShape),
    updates = Reshape(input, flattenedShape),
    axis = 0
)
```
